database/redis.go

```diff
@@ -0,0 +1,74 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package database
+
+import (
+	"context"
+	"strings"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/redis/go-redis/v9"
+	"go.uber.org/zap"
+)
+
+var _ RelayerDatabase = &RedisDatabase{}
+
+type RedisDatabase struct {
+	logger logging.Logger
+	client *redis.Client
+}
+
+func NewRedisDatabase(logger logging.Logger, redisURL string, relayerIDs []RelayerID) (*RedisDatabase, error) {
+	opts, err := redis.ParseURL(redisURL)
+	if err != nil {
+		logger.Error("Failed to parse Redis URL", zap.Error(err), zap.String("url", redisURL))
+		return nil, err
+	}
+
+	// Create a new Redis client.
+	// The server address, password, db index, and protocol version are extracted from the URL
+	// Request timeouts use the default value of 3 seconds
+	client := redis.NewClient(opts)
+	return &RedisDatabase{
+		logger: logger,
+		client: client,
+	}, nil
+}
+
+func (r *RedisDatabase) Get(relayerID common.Hash, key DataKey) ([]byte, error) {
+	ctx := context.Background()
+	compositeKey := constructCompositeKey(relayerID, key)
+	val, err := r.client.Get(ctx, compositeKey).Result()
+	if err != nil {
+		r.logger.Debug("Error retrieving key from Redis",
+			zap.String("key", compositeKey),
+			zap.Error(err))
+		if err == redis.Nil {
```

Comment:
why do we need to return a different error here specifically?

Reply:
We have branching logic depending on if the DB returned a key not found error or some other error:
https://github.com/ava-labs/awm-relayer/blob/af25ff13915796cb208044215c9c5c33edf4f9bc/database/database.go#L49
https://github.com/ava-labs/awm-relayer/blob/acb4428b365176b5b89d003b054e66e29fccbd13/relayer/message_relayer.go#L603